iwyu_regex.cc

```diff
@@ -0,0 +1,47 @@
+//===--- iwyu_regex.cc - iwyu regex implementation ------------------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iwyu_regex.h"
+
+#include <regex>
+#include "llvm/Support/Regex.h"
+
+#include "iwyu_port.h"
+
+namespace include_what_you_use {
+
+bool ParseRegexDialect(const char* str, RegexDialect* dialect) {
+  if (strcmp(str, "llvm") == 0) {
+    *dialect = RegexDialect::LLVM;
+    return true;
+  } else if (strcmp(str, "ecmascript") == 0) {
+    *dialect = RegexDialect::ECMAScript;
+    return true;
+  }
+  return false;
+}
+
+// Returns true if str matches regular expression pattern for the given dialect.
+bool RegexMatch(RegexDialect dialect, const std::string& str,
+                const std::string& pattern) {
+  switch (dialect) {
+    case RegexDialect::LLVM: {
+      llvm::Regex r(pattern);
+      return r.match(str);
+    }
+
+    case RegexDialect::ECMAScript: {
+      std::regex r(pattern, std::regex_constants::ECMAScript);
```

Comment:
I wasn't aware it supports different grammars.

https://en.cppreference.com/w/cpp/regex/basic_regex

Does using `optimize` increase the performance? Can we use `basic` instead of using the LLVM implementation?

Reply:
https://www.regular-expressions.info/stdregex.html

This goes a bit more into detail what the different grammars do.

Reply:
With this patch unchanged:
```
$ cat tests/981/t.cc 
#include <string>

$ time include-what-you-use -Xiwyu --regex=ecmascript -Xiwyu --mapping_file=qt5_11.imp tests/981/t.cc 

tests/981/t.cc should add these lines:

tests/981/t.cc should remove these lines:
- #include <string>  // lines 1-1

The full include-list for tests/981/t.cc:
---

real    0m5,865s
user    0m5,860s
sys     0m0,004s

$ time include-what-you-use -Xiwyu --regex=llvm -Xiwyu --mapping_file=qt5_11.imp tests/981/t.cc 

tests/981/t.cc should add these lines:

tests/981/t.cc should remove these lines:
- #include <string>  // lines 1-1

The full include-list for tests/981/t.cc:
---

real    0m0,425s
user    0m0,417s
sys     0m0,008s
```
Then I applied the following diff, built and ran again:
```
$ git diff
diff --git a/iwyu_regex.cc b/iwyu_regex.cc
index dca92be..c469fe5 100644
--- a/iwyu_regex.cc
+++ b/iwyu_regex.cc
@@ -32,8 +32,10 @@ bool RegexMatch(RegexDialect dialect, const std::string& str,
                 const std::string& pattern) {
   switch (dialect) {
     case RegexDialect::LLVM: {
-      llvm::Regex r(pattern);
-      return r.match(str);
+      std::regex r(pattern, std::regex_constants::extended);
+      return std::regex_match(str, r);
+      // llvm::Regex r(pattern);
+      // return r.match(str);
     }
 
     case RegexDialect::ECMAScript: {

$ time include-what-you-use -Xiwyu --regex=llvm -Xiwyu --mapping_file=qt5_11.imp tests/981/t.cc 

tests/981/t.cc should add these lines:

tests/981/t.cc should remove these lines:
- #include <string>  // lines 1-1

The full include-list for tests/981/t.cc:
---

real    0m5,986s
user    0m5,952s
sys     0m0,012s
```
(note `llvm` is now `std::regex::extended`, LLVM's Regex supports extended regex, so this is a fair comparison)

Just to make sure, I tested with `basic` too:
```
$ git diff
diff --git a/iwyu_regex.cc b/iwyu_regex.cc
index dca92be..5728831 100644
--- a/iwyu_regex.cc
+++ b/iwyu_regex.cc
@@ -32,8 +32,10 @@ bool RegexMatch(RegexDialect dialect, const std::string& str,
                 const std::string& pattern) {
   switch (dialect) {
     case RegexDialect::LLVM: {
-      llvm::Regex r(pattern);
-      return r.match(str);
+      std::regex r(pattern, std::regex_constants::basic);
+      return std::regex_match(str, r);
+      // llvm::Regex r(pattern);
+      // return r.match(str);
     }
 
     case RegexDialect::ECMAScript: {

$ time include-what-you-use -Xiwyu --regex=llvm -Xiwyu --mapping_file=qt5_11.imp tests/981/t.cc 

tests/981/t.cc should add these lines:

tests/981/t.cc should remove these lines:
- #include <string>  // lines 1-1

The full include-list for tests/981/t.cc:
---

real    0m5,930s
user    0m5,910s
sys     0m0,016s
```

So `std::regex` seems objectively worse. :-)

Reply:
I would expect `optimize` to actually make things worse, since we create a new `regex` object every time we evaluate a pattern -- `optimize` trades construction time for evaluation time.

I tested it both with `basic` and `ECMAScript`, but could not observe any difference in runtime.

Reply:
(I have a pretty cool experiment here with a runtime-polymorphic regex class where we could potentially cache the compiled regex to make some gains, but I wanted to start with this simpler variation first. Take a look here if you're interested: https://github.com/include-what-you-use/include-what-you-use/compare/master...kimgr:include-what-you-use:polyregex-variant)

Reply:
> I would expect `optimize` to actually make things worse, since we create a new `regex` object every time we evaluate a pattern -- `optimize` trades construction time for evaluation time.
> 

☹
Just for giggles give it a spin (since I can't). The evaluation time is bad - maybe the construction time is not as bad. Unfortunately I see no `constexpr` anywhere in the signatures.

Reply:
> (I have a pretty cool experiment here with a runtime-polymorphic regex class where we could potentially cache the compiled regex to make some gains, but I wanted to start with this simpler variation first. Take a look here if you're interested: [master...kimgr:include-what-you-use:polyregex-variant](https://github.com/include-what-you-use/include-what-you-use/compare/master...kimgr:include-what-you-use:polyregex-variant))

Yeah - I don't get it 🥴. Some patterns I haven't used yet and modern C++ are things that make my head hurt.

Reply:
I already did give it a spin, as mentioned up-thread
> I tested it both with basic and ECMAScript, but could not observe any difference in runtime.

Reply:
> I already did give it a spin, as mentioned up-thread
> 
> > I tested it both with basic and ECMAScript, but could not observe any difference in runtime.

Ah, sorry. I misunderstood that line.